src/Microsoft.Health.ControlPlane.CosmosDb/Features/Storage/ControlPlaneDataStore.cs

```diff
@@ -106,41 +140,92 @@ private async Task<T> GetSystemDocumentByIdAsync<T>(string id, string partitionK
+                return new UpsertResponse<T>((T)(dynamic)response.Resource, outcomeType, response.Resource.ETag);
+            }
+            catch (DocumentClientException dce)
+            {
+                if (dce.StatusCode == HttpStatusCode.TooManyRequests)
                 {
-                    _logger.LogError($"{typeof(T)} with id {id} was not found in Cosmos DB.");
+                    throw new RequestRateExceededException(dce.RetryAfter);
                 }
 
-                return result;
+                _logger.LogError(dce, "Unhandled Document Client Exception");
+                throw;
             }
         }
 
-        private async Task<T> UpsertSystemObjectAsync<T>(T systemObject, string partitionKey, CancellationToken cancellationToken)
-            where T : class
+        private static RequestOptions GetRequestOptions(string partitionKey, string eTag)
         {
-            EnsureArg.IsNotNull(systemObject, nameof(systemObject));
             var eTagAccessCondition = new AccessCondition();
 
+            if (!string.IsNullOrWhiteSpace(eTag))
+            {
+                eTagAccessCondition.Type = AccessConditionType.IfMatch;
+                eTagAccessCondition.Condition = eTag;
+            }
+
             var requestOptions = new RequestOptions
             {
                 PartitionKey = new PartitionKey(partitionKey),
                 AccessCondition = eTagAccessCondition,
             };
+            return requestOptions;
+        }
+
+        private async Task DeleteSystemDocumentsByIdAsync<T>(string id, string partitionKey, string eTag, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(id, nameof(id));
+            EnsureArg.IsNotNullOrEmpty(partitionKey, nameof(partitionKey));
 
             try
             {
-                ResourceResponse<Document> response = await _documentClient.Value.UpsertDocumentAsync(
-                    _collectionUri,
-                    systemObject,
-                    requestOptions,
-                    true,
+                _logger.LogDebug($"Obliterating {id}");
+
+                StoredProcedureResponse<IList<string>> response = await _retryExceptionPolicyFactory.CreateRetryPolicy().ExecuteAsync(
+                    async () => await _hardDeleteIdentityProvider.Execute(
+                        _documentClient.Value,
+                        _collectionUri,
+                        id,
+                        eTag),
                     cancellationToken);
-                _logger.LogInformation("Request charge: {RequestCharge}, latency: {RequestLatency}", response.RequestCharge, response.RequestLatency);
 
-                return (T)(dynamic)response.Resource;
+                _logger.LogDebug($"Hard-deleted {response.Response.Count} documents, which consumed {response.RequestCharge} RUs. The list of hard-deleted documents: {string.Join(", ", response.Response)}.");
```

Comment:
I don't think this function can currently delete more than one document since it is happening by id.

Reply:
This is a generic method, I will leave it as is.

Reply:
There can only be one document if the id and partition key is specified. Let's not call it "documents".